opal/mca/accelerator/cuda/accelerator_cuda.c

```diff
@@ -0,0 +1,775 @@
+    int (*cuCtxSetCurrent)(CUcontext);
+    int (*cuStreamSynchronize)(CUstream);
+    int (*cuStreamDestroy)(CUstream);
+#if OPAL_CUDA_GET_ATTRIBUTES
+    int (*cuPointerGetAttributes)(unsigned int, CUpointer_attribute *, void **, CUdeviceptr);
+#endif /* OPAL_CUDA_GET_ATTRIBUTES */
+};
+typedef struct accelerator_cuda_func_table  accelerator_cuda_func_table_t;
+
+static accelerator_cuda_func_table_t accelerator_cuda_func;
+
+opal_accelerator_base_module_t opal_accelerator_cuda_module =
+{
+    accelerator_cuda_check_addr,
+    accelerator_cuda_memcpy,
+    accelerator_cuda_vector_memcpy,
+    accelerator_cuda_memmove,
+    accelerator_cuda_malloc,
+    accelerator_cuda_free,
+    accelerator_cuda_get_address_range,
+
+    accelerator_cuda_is_ipc_enabled,
+    accelerator_cuda_get_handle_size,
+    accelerator_cuda_get_handle,
+    accelerator_cuda_open_handle,
+    accelerator_cuda_close_handle,
+
+    accelerator_cuda_host_register,
+    accelerator_cuda_host_unregister,
+
+    accelerator_cuda_get_device,
+    accelerator_cuda_device_can_access_peer
+};
+
+static CUstream memcpyStream = NULL;
+static bool stage_two_init_complete = false;
+static opal_mutex_t accelerator_cuda_init_lock;
+/* Unused variable that we register at init time and unregister at fini time.
+ * This is used to detect if user has done a device reset prior to MPI_Finalize.
+ * This is a workaround to avoid SEGVs.
+ */
+static int checkmem;
+static int ctx_ok = 1;
+
+static int accelerator_cuda_populate_func_table(opal_dl_handle_t *libcuda_handle)
+{
+    /* Map in the functions that we need.  Note that if there is an error
+     * the macro OPAL_CUDA_DLSYM will print an error and call return.  */
+    OPAL_CUDA_DLSYM(libcuda_handle, cuStreamCreate);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuCtxGetCurrent);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuMemHostRegister);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuMemHostUnregister);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuPointerGetAttribute);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuMemcpyAsync);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuMemcpy);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuMemcpy2D);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuMemFree);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuMemAlloc);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuMemGetAddressRange);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuIpcOpenMemHandle);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuIpcCloseMemHandle);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuIpcGetMemHandle);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuCtxGetDevice);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuDeviceCanAccessPeer);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuCtxSetCurrent);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuStreamSynchronize);
+    OPAL_CUDA_DLSYM(libcuda_handle, cuStreamDestroy);
+#if OPAL_CUDA_GET_ATTRIBUTES
+    OPAL_CUDA_DLSYM(libcuda_handle, cuPointerGetAttributes);
+#endif /* OPAL_CUDA_GET_ATTRIBUTES */
+    return OPAL_SUCCESS;
+}
+
+opal_accelerator_base_module_t* accelerator_cuda_internal_init(void)
```

Comment:
This init function should fail if:

1. The user did not initialize a CUDA context before this function got called
2. There is not a CUDA device present on the instance

I think we need to make a call into the CUDA API in this method to confirm that the CUDA component we are returning is functional.  If it is not functional, we should return the Null Accelerator to the user.